ethereum: Add property tests and instructions for running them with KEVM

lucasmt opened the pull request:
* Add new Forge property tests for the Ethereum core messaging contracts in `forge-std`.
* Add wrappers for the `Getters`, `GovernanceStructs`, `Setters` and `Setup` tests to be executed symbolically by the [KEVM formal verification framework](https://github.com/runtimeverification/evm-semantics).
* Add `rv-helpers` folder with helper contracts for property tests.
* Add `run-kevm.sh` script for reproducing KEVM proofs.
* Add `PROOFS.md` with instructions for installing and running KEVM to reproduce the proofs.
* Add `wormhole-lemmas.k` file with helper lemmas used in the KEVM proofs.
* Edit `Makefile` to ignore the KEVM versions of the tests when running `forge test`.

SEJeff reviewed (commented):
This is a fantastic PR, but I do have a few thoughts I'd like to share:

Getters.sol
-----------

1. no coverage for Getters.getGuardianSetExpiry(). Was this intentional?


GovernanceStructs.sol
---------------------

1. The length checks in the following functions don't have the positive and negative cases tested:
  a. parseContractUpgrade()
  b. parseGuardianSetUpgrade()
  c. parseSetMessageFee()
  d. parseTransferFees()
  
  I'm guessing the fuzzer is unlikely to get both and successfully get to this check without hitting a prior revert.


Messages.sol
------------

1. parseAndVerifyVM() has some missing negative cases for the following:
  a. wrong guardian set index or expired guardian set: `return (false, "guardian set has expired");`
  b. signature length less than quorum of the guardian set `return (false, "no quorum");`
  c. invalid signature `return (false, invalidReason);`
2. verifySignatures() ensure that the signature indices are ascending
3. parseVM() either the positive or negative case for the vm version being incompatible


Setup.sol
---------

Running `forge coverage ...` shows zero coverage for Setup.sol, but it isn't able to understand you're calling it via the proxy unfortunately. I guess this is a noop and we can find another way to cover this (limitations of `forge coverage` I suppose).

lucasmt commented on ethereum/forge-test/GovernanceStructs.t.sol:
```diff
@@ -0,0 +1,320 @@
+        testParseTransferFeesWrongAction(module, action, chain, amount, recipient);
+    }
+
+    // Needs loop invariant for unbounded bytes type
```
The main goal of this engagement was to formally verify Wormhole functionality, by using our tool KEVM to symbolically execute property tests. But we decided to limit the scope of this engagement to those tests that didn't include loops, and leave tests that did include loops to a follow-up engagement. This is because symbolically executing loops requires setting up a loop invariant, which requires more work than straight-line code. The `...InvalidSize` tests actually have a hidden loop because the parameter of type `bytes` (of arbitrary length) needs to be copied into memory. So they were left out of the present scope because they would need a loop invariant to be formally verified.

lucasmt commented on ethereum/forge-test/GovernanceStructs.t.sol:
```diff
@@ -0,0 +1,320 @@
+        vm.expectRevert();
+
+        gs.parseContractUpgrade(encodedUpgrade);
+    }
```
We wrote these tests with symbolic execution using our KEVM tool in mind, and in that case this would not be an issue because, unlike fuzzing, symbolic execution covers every possible code path. However, we ended up not actually running symbolic execution on the `...InvalidSize` tests in this engagement (see my comment [here](https://github.com/wormhole-foundation/wormhole/pull/2956#discussion_r1240580012)), so based on your suggestions I have refactored the tests so they work better as fuzz tests.

lucasmt commented on ethereum/forge-test/MessagesRV.t.sol:
```diff
@@ -0,0 +1,214 @@
+// test/Messages.sol
+// SPDX-License-Identifier: Apache 2
```
`Messages.t.sol` was a pre-existing test contract. `MessagesRV.t.sol` are tests that we wrote during this engagement to try to generalize the pre-existing tests. We placed them in a separate file to differentiate the two.

lucasmt commented on ethereum/forge-test/Setup.t.sol:
```diff
@@ -0,0 +1,117 @@
+
+        address[] memory keys = new address[](1);
+        keys[0] = 0xbeFA429d57cD18b7F8A4d91A2da9AB4AF05d0FBe;
+        //keys[0] = vm.addr(testGuardian);
```
Apologies, this was an oversight, but I moved it to the previous line as you suggested because I think the context is helpful.

lucasmt commented:
@SEJeff I've responded to your comments and accepted all your suggestions, thank you.

>  1. no coverage for Getters.getGuardianSetExpiry(). Was this intentional?

Yes, during the engagement we determined that `guardianSetExpiry` was unused ([the `expireGuardianSet` function](https://github.com/wormhole-foundation/wormhole-rv/blob/f9758b38e7efd9d7e5b21a48e009c65e2138710c/ethereum/contracts/Setters.sol#L13) instead uses a constant as the expiration interval). `getGuardianSetExpiry` is the only function that uses it, and there's no way to write to it. [Etherscan](https://etherscan.io/address/0x98f3c9e6E3fAce36bAAd05FE09d375Ef1464288B#readProxyContract#F5) shows this variable is set to 0.

>  1. The length checks in the following functions don't have the positive and negative cases tested:
>        a. parseContractUpgrade()
>        b. parseGuardianSetUpgrade()
>        c. parseSetMessageFee()
>        d. parseTransferFees()
> 
> 
> I'm guessing the fuzzer is unlikely to get both and successfully get to this check without hitting a prior revert.

I think the case where it passes the length check is exercised in `testParseContractUpgrade()` and co., no? Regarding the negative case, see my answers to your comments.

>  1. parseAndVerifyVM() has some missing negative cases for the following:
>        a. wrong guardian set index or expired guardian set: `return (false, "guardian set has expired");`
>        b. signature length less than quorum of the guardian set `return (false, "no quorum");`
>        c. invalid signature `return (false, invalidReason);`
> 
>  2. verifySignatures() ensure that the signature indices are ascending
> 
>  3. parseVM() either the positive or negative case for the vm version being incompatible

Our focus on `MessagesRV.t.sol` was to make the previously-written tests in `Messages.t.sol` more general, for example by considering an arbitrary number of guardians. Because of that, we only covered those behaviors that were already being tested in `Messages.t.sol`.

> Running `forge coverage ...` shows zero coverage for Setup.sol, but it isn't able to understand you're calling it via the proxy unfortunately. I guess this is a noop and we can find another way to cover this (limitations of `forge coverage` I suppose).

That is unfortunate, I don't know if there is a way to get around this limitation.

SEJeff commented:
Hey @lucasmt I added 61e793c9b3e71a7893516f7295b79d44048ff2d5 and 74fd7b15397c352c79a34b5be7bde62c3ec519f6 to make it run, but one of the tests is still failing:

```bash
Running 4 tests for forge-test/MessagesRV.t.sol:TestMessagesRV
[PASS] testCannotVerifySignaturesWithInvalidSignature1(bytes,(uint256[],uint8,uint32),(bytes32,bytes32,uint8,uint8)) (runs: 256, μ: 58681, ~: 48952)
[PASS] testCannotVerifySignaturesWithInvalidSignature2(bytes,(uint256[],uint8,uint32),uint8,uint256) (runs: 256, μ: 61299, ~: 53286)
[FAIL. Reason: Arithmetic over/underflow Counterexample: calldata=0xd4dec08c000000000000000000000000000000000000000000000000000000000000008000000000000000000000000000000000000000000000000000000000000000c000
00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000010000000000000000
00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000600000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000
00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000, args=[0x00, ([], 0, 0), 0, 0]] testCannotVerifySignaturesWithOutOfBoundsSignature(bytes,(uint256[],uint8,u
int32),uint8,uint8) (runs: 1, μ: 46327, ~: 46327)
[PASS] testVerifySignatures(bytes,(uint256[],uint8,uint32)) (runs: 256, μ: 44389, ~: 33308)
Test result: FAILED. 3 passed; 1 failed; 0 skipped; finished in 6.33s
```

lucasmt commented:
@SEJeff sorry about that, fixed now. The change from `vm.assume` to `bound` introduced a possible arithmetic underflow on the expression `params.guardianCount - 1`. I've fixed it by adding an `assume` that `params.guardianCount > 0`.

tbjump reviewed (approved):
more tests can't hurt
